src/Domain/Services/CheckoutFields.php

```diff
@@ -0,0 +1,653 @@
+	}
+
+	/**
+	 * Returns an array of fields keys for the contact group.
+	 *
+	 * @return array An array of fields keys.
+	 */
+	public function get_contact_fields_keys() {
+		return $this->fields_locations['contact'];
+	}
+
+	/**
+	 * Returns an array of fields keys for the additional area group.
+	 *
+	 * @return array An array of fields keys.
+	 */
+	public function get_additional_fields_keys() {
+		return $this->fields_locations['additional'];
+	}
+
+	/**
+	 * Returns an array of fields for a given group.
+	 *
+	 * @param string $location The location to get fields for (address|contact|additional).
+	 *
+	 * @return array An array of fields.
+	 */
+	public function get_fields_for_location( $location ) {
+		if ( in_array( $location, array_keys( $this->fields_locations ), true ) ) {
+			return $this->fields_locations[ $location ];
+		}
+	}
+
+	/**
+	 * Validates a field value for a given group.
+	 *
+	 * @param string $key The field key.
+	 * @param mixed  $value The field value.
+	 * @param string $location The location to validate the field for (address|contact|additional).
+	 *
+	 * TODO: we might not need the location param here.
+	 *
+	 * @return true|\WP_Error True if the field is valid, a WP_Error otherwise.
+	 */
+	public function validate_field_for_location( $key, $value, $location ) {
+		if ( ! $this->is_field( $key ) ) {
+			// translators: %s field key.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_invalid', \sprintf( __( 'The field %s is invalid.', 'woo-gutenberg-products-block' ), $key ) );
+		}
+
+		if ( ! in_array( $key, $this->fields_locations[ $location ], true ) ) {
+			// translators: %1$s field key, %2$s location.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_invalid_location', \sprintf( __( 'The field %1$s is invalid for the location %2$s.', 'woo-gutenberg-products-block' ), $key, $location ) );
+		}
+
+		$field = $this->additional_fields[ $key ];
+		if ( ! empty( $field['required'] ) && empty( $value ) ) {
+			// translators: %s field key.
+			return new \WP_Error( 'woocommerce_blocks_checkout_field_required', \sprintf( __( 'The field %s is required.', 'woo-gutenberg-products-block' ), $key ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns true if the given key is a valid field.
+	 *
+	 * @param string $key The field key.
+	 *
+	 * @return bool True if the field is valid, false otherwise.
+	 */
+	public function is_field( $key ) {
+		return array_key_exists( $key, $this->additional_fields );
+	}
+
+	/**
+	 * Persists a field value for a given order. This would also optionally set the field value on the customer.
+	 *
+	 * @param string    $key The field key.
+	 * @param mixed     $value The field value.
+	 * @param \WC_Order $order The order to persist the field for.
+	 * @param bool      $set_customer Whether to set the field value on the customer or not.
+	 *
+	 * @return void
+	 */
+	public function persist_field_for_order( $key, $value, $order, $set_customer = true ) {
```

Comment:
This function would set a value for an order and optionally set it for the customer, it's used when placing the order.